src/System.Slices/System/Buffers/Memory2.cs

```diff
@@ -0,0 +1,113 @@
+﻿using System.Buffers;
+using System.Runtime.CompilerServices;
+using System.Threading;
+
+namespace System
+{
+    public struct Memory2<T> 
+    {
+        MemoryManager<T> _manager;
+        long _id;
+        int _index;
+        int _length;
+
+        internal Memory2(MemoryManager<T> manager, long id)
+            : this(manager, id, 0, manager.GetSpan(id).Length)
+        { }
+
+        private Memory2(MemoryManager<T> manager, long id, int index, int length)
+        {
+            _manager = manager;
+            _id = id;
+            _index = index;
+            _length = length;
+        }
+
+        public static Memory2<T> Empty => EmptyManager.Shared.Memory;
+
+        public int Length => _length;
+
+        public bool IsEmpty => Length == 0;
+
+        public Memory2<T> Slice(int index)
+        {
+            return new Memory2<T>(_manager, _id, _index + index, _length - index);
+        }
+        public Memory2<T> Slice(int index, int length)
+        {
+            return new Memory2<T>(_manager, _id, _index + index, length);
+        }
+
+        public Span<T> Span => _manager.GetSpan(_id).Slice(_index, _length);
+
+        public DisposableReservation Reserve() => new DisposableReservation(_manager, _id);
```

Comment:
This will need to be updated to match the language extensions once they are designed. I would expect that the Span property and Reserve method will become just one thing; and DisposableReservation won't exist at all.